### ISSUE TITLE
Fix delay UX

### ISSUE THREAD

**Issue by Timmmm** [linux-pam/linux-pam#778]
PAM's `pam_unix` has a default fixed delay of approximately 2 seconds. As far as I can tell most software and distros stick with this and use it for everything - ssh, sudo, login, etc.

From a usability point of view that sucks. It's a significant amount of time to wait when you make a typo and it add dubious security benefits. There are no security benefits at all for sudo (you can just use `unix_chkpwd`), and I'm dubious about `sshd` - I expect you can get around it just by creating a load of parallel connections (though I haven't tried it).

I would like to fix this annoying paper cut. I can think of the following solutions and wanted your feedback on what you would accept: 

1. Just change the default delay to 0.5s. This is short enough to be not annoying to users making typos but isn't really going to change any security properties (nobody is going to care about waiting 4x longer to brute force a login). This also has the advantage of being really really easy to do.
2. Add a parameter to `faildelay` to have increasing delays, and store the current delay count as part of the session (if that is possible; I couldn't quite figure it out and it seems like a bad idea to use global variables). This isn't much work either.
3. Add a parameter to `faildelay` to have increasing delays, and persist the fail information to disk in the same way that `faillock` does. This is a ton of work. I'll probably give up if this is the only acceptable solution.
4. Add a parameter to `faillock` to have increasing delays.
5. Add a way to *decrease* the delay. The issue with API at the moment is that modules in the PAM stack can only increase the delay. The way Debian has the configs set up, there's a default config with the 2s delay, which means you can't e.g. force it back to 0 for `sudo`. It's not ideal to have the default delay as 0 because that isn't fail-safe. Every config *except* `sudo` then has to remember to set the delay.

My preference is 1. What do you think?

**Comment by Teerath-Agarwal** [comment]
I was annoyed by the delay and was thinking opening a PR today itself, when I came across this issue. I support your idea more or less, but I think, we can keep a count of number of times the password has been incorrectly entered consecutively. If its less than three, then a 0.1 second delay is enough. Otherwise, (0.5*count) second delay. Doesn't that sound better?

**Comment by Timmmm** [comment]
Yes, that is clearly the best way to do it, and it's how most password systems work. I think Windows has done it since at least XP.

However because PAM has a weird architecture (it's a library in the client rather than a single persistent system component) it's quite difficult to store persistent information. The `faillock` plugin does it by writing fail counts (but not times) to disk, but it's quite complex.

So you would either need to make a new plugin with a similar complexity to `faillock`, or patch `faillock`, which doesn't seem ideal because that's not what it's meant for.

Doable but quite a lot of work, whereas changing the default is a one line change. It also wouldn't require distros to change their configs so we might actually get a (suboptimal) fix before I die!

Glad someone else agrees anyway! I'm going to make a PR. If you want to make a PR for a new exponential delay plugin go ahead. I did start writing one but I hate C so...

**Comment by meribold** [comment]
I agree that lowering the default delay would be a considerable UX improvement. While something more elaborate may be even better, I feel that having the simple fix now would be great regardless (it doesn't preclude further improvements in the future). IIRC, I used to lower the delay with a line like `auth optional pam_faildelay.so delay=100000` in `/etc/pam.d/system-auth`, but that doesn't seem to work anymore. Currently the only option seems to be to remove the delay completely.

**Comment by Timmmm** [comment]
I made a PR to fix it - #789 - but it's been waiting for a review for over a year. :-/

**Comment by matezoltanfarkas** [comment]
@ldv-alt what do you think of this?

**Comment by ldv-alt** [comment]
The current 2-second delay is a long-standing default that has been in place for many years. Changing something so fundamental, which affects a vast number of systems and distributions, is a significant undertaking that requires a very strong justification.

While you find the delay annoying, my personal experience—and that of all colleagues I've consulted—is actually the opposite. We find the short pause is just enough to take a breath and refocus after a mistyped password, preventing a second fumbled attempt.

Given these differing experiences, it seems the perception of this delay is quite subjective. To move forward with a change to the default behavior, I would need more than anecdotal feedback. Could you please provide some compelling evidence to support the claim that a 2-second delay is widely considered a poor user experience?
Have you come across any usability studies or research that have explored this and demonstrated a clear user preference for a shorter delay after a mistyped password?
Without stronger data, it's hard to justify changing a decades-old default that so many users are accustomed to.

**Comment by matezoltanfarkas** [comment]
thanks for looking into this @ldv-alt !
I think it all comes down to flexible defaults.
I fully agree with you on changing settings only if it is necessary given the wide adoption of this project; and I don't think the delay is poor user experience either.

Neither do I have data on this and I personally find it hardly feasible to collect it -- or at least I am in no such position whatsoever.

What does contribute to better user experience in general is flexibility and I believe hardcoded values take that away. I think a lot of sysadmins/users/tweakers could benefit from a modifiable timeout argument (with a default of 2s, of course).

...and if stuff breaks, it's on them then!

**Comment by ldv-alt** [comment]
> Interesting, I was expecting pushback on security grounds (which I believe is the motivation for this delay).

The reason behind fail delays is to slow down brute-force attacks.
However, the delay of 2 seconds we're talking about was introduced in the previous century, and the rationale for this specific value is unknown. Nowadays there are additional means available to slow down brute-force attacks more efficiently, e.g. by using `pam_faillock` module and restricting access to privileged helpers, even though the latter is not very popular.

> I never considered that people might _want_ arbitrary delays in their computing experience.

I wouldn't frame it this way.  It's not an arbitrary general delay but a quite predictable delay in a very specific situation.
Apparently, most of users have been experiencing this particular value of the fail delay all the time,
so they got used to it.  For example, when I mistype a password, I use these ~2 seconds to calm down and gain enough concentration to type the password correctly for the second time.

> The problem with changing the configuration is that a) it's actually quite complicated to do, and b) then we have to convince every distro to do it. That's exponentially more work. Also I already tried to get Red Hat to do that and they said the default should be changed here instead.
> 
> So I would definitely rather persuade you that people do prefer a shorter delay. What kind of evidence would convince you? I could do a poll on Reddit/Twitter or something?

As I said earlier, if there any usability studies or research that have explored this and demonstrated a clear user preference for a shorter delay after a mistyped password, it would be great to have a look at their results.
Of course you can do some polling, but how representative that polling would be?

**Comment by Timmmm** [comment]
> Of course you can do some polling, but how representative that polling would be?

If we do it in a neutral place like Reddit or Twitter then I think it would be reasonably representative of "normal users" surely?

I'm not sure what a "user study" would be other than showing users a login interface with a 2 second delay and a 0.5 second delay and asking them which they prefer. Is that what you mean? If so I could probably do that too.

**Comment by felagund** [comment]
This is not scientific, but just putting a string "how to remove the delay after mistyping password sudo" into a search engine gives me a list of these:

https://askubuntu.com/questions/877385/how-can-i-lower-the-delay-after-incorrectly-entered-login-and-sudo-passwords
https://www.baeldung.com/linux/wrong-password-delay
https://discussion.fedoraproject.org/t/how-to-reduce-the-delay-after-mistype-my-password-when-running-sudo/162149
https://biancatamayo.me/blog/why-do-incorrect-sudo-passwords-take-so-long/
https://unix.stackexchange.com/questions/658771/remove-delay-after-wrong-sudo-password-entry-on-arch-linux
https://superuser.com/questions/165550/change-password-timeout-on-linux
https://coderwall.com/p/8_hxog/removing-the-delay-when-entering-an-incorrect-sudo-password
https://www.reddit.com/r/archlinux/comments/pmdv3j/how_to_deactivate_delay_after_entering_the_wrong/
https://bbs.archlinux.org/viewtopic.php?id=302965
https://forum.manjaro.org/t/remove-delay-after-entering-a-wrong-password/162617
https://askubuntu.com/questions/1245693/change-the-delay-that-occurs-after-entering-an-incorrect-password-in-the-gdm3-lo
https://unix.stackexchange.com/questions/2126/why-is-there-a-big-delay-after-entering-a-wrong-password
https://bbs.archlinux.org/viewtopic.php?id=268668
https://bbs.archlinux.org/viewtopic.php?id=277348
https://unix.stackexchange.com/questions/40954/how-does-one-change-the-delay-that-occurs-after-entering-an-incorrect-password
https://unix.stackexchange.com/questions/122422/is-it-possible-to-remove-the-delay-on-wrong-password
https://unix.stackexchange.com/questions/615783/how-to-lower-the-delay-after-typing-wrong-password-in-sudo-command-in-arch-linux
https://unix.stackexchange.com/questions/176375/password-fail-delay-is-too-long-linux-mint-17

Given that most people do not change defaults at all, and even less people ask about it on the internet, this is clearly something that is important for a  quite a few people. It is hard to know for sure, of course.

However, there does not seem to be a good rationale for the current state of things. Personally for me, if anything, I lose concentration during the two seconds I wait for the next attempt. I guess this varies. Anyway, if the only reasons for the delay are:
1) it has been always this way
2) it provides some with time to concentrate
3) users got used to this
I think the rationale is quite poor. Also, what would be the downside of changing this? I doubt many people will notice that the delay is gone, and if they do, I think it will be even rarer for them to be sad about it. This is not a change that requires anybody to change their workflow. Nothing stops people for concentrating for two seconds before typing the password again. But why push it to everybody?

**Comment by felagund** [comment]
I would personally prefer solution number 2 (increasing delays). One of the links I posted says that the delay is there so that an attacker cannot measure the time it took the system to refute a bad password. possibly making brute-force attack easier (by knowing how long the password is). No idea if this is actually relevant or not here. If yes, I am kind of doubt that 2 seconds are not a massive overkill, given how fast computers are nowadays.

**Comment by Timmmm** [comment]
> so that an attacker cannot measure the time it took the system to refute a bad password

Yes the delay is randomised for this purpose so it isn't actually always exactly 2 seconds (or 0.5) but a random amount more or less than that.

> I would personally prefer solution number 2 (increasing delays)

I agree, but that's actually a lot more complex to implement, partly because of PAM's slightly odd architecture, but also for things like SSH where an attacker could use a botnet and start from 0 delay many times.

> Nothing stops people for concentrating for two seconds before typing the password again.

Agreed. We need to know what would convince @ldv-alt . I can do some research but I really don't want to put a lot of effort into polling people only for the results to be disregarded for some reason.

**Comment by felagund** [comment]
> What we dont see of course is all the people that just love the two second delay, as they, presumably, dont post in forums about how to keep the value they are happy with.

Is there anybody who _loves_ it? Even in this thread. Do you really _love_ the delay? I assume most people do not care as they do not think about changing how their computers work. There is for sure a group of people who are not happy about the delay. "Hate" would probably be too strong a word.



> About defaults: if there is no correct, good or bad default, the default cannot be bad.
This is misleading. Good default does not mean that everybody agrees on it. Good default is a default that 1) the most people prefer and does not cause 2) undue strain on those who do not prefer it.

1) @Timmmm is asking for a confirmation on what kind of user survey would be acceptable as valud argument for what people actually prefer. It would be good if people who actually decide, that is maintainers linux_pam, responded to this request.

2) For 2), an example of an undue strain of lower delay is needed. I cannot imagine one, nor was one provided, but maybe there is?



> Everybody I asked about this told me, 2 seconds are good and right and they don't understand why you want to change that to 0.5 seconds.

Can I ask who those people are and how many are there and what their reasons are? Also, how come they do not understand why anybody would want to change. What is mysterious about not wanting to wait needlessly? I can imagine some people do not care, but is it incomprehensible that some people do? I always assumed the delay was there to guard against brute force attacks. Now that I see that the reason is "it has always been that way", I do not understand the delay at all.

Please note the reasons I listed above that seem to have been stated as reasons for the delay:
> * it has been always this way
> * it provides some with time to concentrate
> * users got used to this

Are there any other reasons why you or the people you asked prefer the delay? I actually really struggle to understand why the delay is valuable. I can understand that some people are opposed to change on principle (it is called conservatism) and I do agree that needless change has dubious benefits and that all changes tend to have some costs. However, what are the costs here? Nobody came up with any so far. The benefits are clear - time saved. On individual level, it is a few seconds. Given the number of linux users, I suspect on global scale we are talking about human-years wasted on this delay. (assuming there are ten milion Linux users who mistype their password at least three times a year. I for sure count as at least a hundred :-))

**Comment by Timmmm** [comment]
> For this reason, it's configurable! Disable the delay in pam_unix and add pam_faildelay with the delay you would like to see.

I just wanted to revisit the idea that users can configure this themselves, because you make it sound easy whereas in practice almost nobody will be able to make this configuration change before giving up. Let's look at how you do this. I'm using Rocky 9.

1. Magically know that PAM is the thing you need to configure. No idea how I learnt this. You can probably google this I guess.
2. Find `/etc/pam.d` - not too hard.
3. Ok there are 44 files in here. Do I want `login`? `passwd`? `password-auth`? `system-auth`? Let's try `password-auth`.

```
# Generated by authselect on Fri Jul 25 10:05:26 2025
# Do not modify this file manually.

auth        required                                     pam_env.so
auth        required                                     pam_faildelay.so delay=2000000
auth        [default=1 ignore=ignore success=ok]         pam_usertype.so isregular
auth        [default=1 ignore=ignore success=ok]         pam_localuser.so
auth        sufficient                                   pam_unix.so nullok
auth        [default=1 ignore=ignore success=ok]         pam_usertype.so isregular
auth        sufficient                                   pam_sss.so forward_pass
auth        required                                     pam_deny.so
...
```

Ok so it's not a standard syntax file (JSON, TOML, etc.). And I'm not meant to actually edit this file. Now I have to learn what `authselect` is.

Also note how they added `pam_faildelay.so delay=2000000`, presumably because `pam_ss.so` *doesn't* have a default delay? So the `pam_unix.so` delay is completely pointless anyway (in realistic configs).

4. `man authselect`

> If the provided profile set is not sufficient, the administrator may create a custom profile by putting it in a special profile directory (/etc/authselect/custom). By doing so, the profile is immediately usable by authselect. See authselect-profiles(5) for more information on  extending existing profiles.

5. `man authselect-profiles` (argh I just wanted to change a delay! why am I here?)

> To register a new profile within authselect, create a directory in one of the authselect profile locations with the files listed above. Not all of the files must be present, only README is mandatory. Other files can be created on per-need basis.
>
> You may find authselect create-profile command helpful when creating new profile. See authselect(8) manual page or authselect create-profile --help for more information.

6. Ok that sounds useful, I got to `sudo authselect create-profile mine --base-on-default` without too much trouble.

7. Ok I open `/etc/authselect/custom/mine/password-auth` but it's empty. Hmm I thought `--base-on-default` would copy the existing one... Maybe I'm not using the default profile (even though I never changed it).

8. Aha!

```
$ sudo authselect current
Profile ID: sssd
```

What is `sssd`?? Oh well let me try again:

```
$ sudo authselect create-profile mine --base-on sssd
...
$ sudo rm -rf /etc/authselect/custom/mine
$ sudo authselect create-profile mine --base-on sssd
```

9. Ok now it isn't empty. There's a tantalising `pam_faildelay.so delay=2000000`, let's try changing that to `pam_faildelay.so delay=500000`.

10. And then I guess I have to do

```
$ sudo authselect select mine
[error] Unable to get profile information [2]: No such file or directory
```

Oh it's `custom/mine`

11. Ok it didn't break anything but there's still a 2s delay.

12. If I search for `delay` it's also in `system-auth`. I'll try that too.

13. Nope. Maybe I need to refresh the config somehow?

[Unclear if this is really required](https://unix.stackexchange.com/questions/363978/how-to-restart-pam-d-service-after-changing-of-common-account-pc-or-common-auth) - my understanding of the PAM architecture is that the top answer is correct; there's no PAM daemon and you don't need to reload anything. I also don't seen to have this `pam-auth-update` that's mentioned, but I guess they aren't making it up so what is that? Oh well.

14. I actually got this far in the past, and at this point I had to resort to reading the PAM source code - which took quite a while - whereupon I discovered the built in 2s minimum. So again you have to magically know that you have to add `nodelay` to `pam_unix`. But where?

```
$ rg pam_unix
password-auth
9:auth        sufficient                                   pam_unix.so {if not "without-nullok":nullok}
18:account     required                                     pam_unix.so
27:password    sufficient                                   pam_unix.so sha512 shadow {if not "without-nullok":nullok} use_authtok
37:session     required                                     pam_unix.so

smartcard-auth
13:account     required                                     pam_unix.so
24:session     required                                     pam_unix.so

system-auth
14:auth        sufficient                                   pam_unix.so {if not "without-nullok":nullok}
25:account     required                                     pam_unix.so
34:password    sufficient                                   pam_unix.so sha512 shadow {if not "without-nullok":nullok} use_authtok
44:session     required                                     pam_unix.so

fingerprint-auth
13:account     required                                     pam_unix.so
26:session     required                                     pam_unix.so
```

It seems like the custom PAM configuration format was not sufficient so we now also have a custom templating system, yeay!

Anyway I guess I now have to learn what `auth`, `account`, `password`, `session`, `required` and `sufficient` mean. I think I did look this up last time but I've forgotten now.

15. Give up and ask ChatGPT what they mean. It gives me something plausible-sounding, including the WTF that `password` is related to password *changes*. I vaguely remember that.

Based on that I think I want `nodelay` on basically all of them (probably not needed for `password` though ironically). And I don't use fingerprints or smartcards so I can ignore those.

16. I add `nodelay` to `system-auth` and `password-auth` on all of the `pam_unix.so`s. Only the `auth` section has `pam_faildelay.so` so maybe I'd want to add it to the other sections? I don't care at this point; I just want something that works.

17. It still doesn't work anyway. Is it even using my config files? I'll add a syntax error to check (after opening a `sudo bash` just in case... this seems very dangerous; I could easily lock myself out).

18. Nope. I added random characters to `system-auth` and `password-auth` and it seems to have just ignored them. Let me check

```
$ ls -l /etc/pam.d/password-auth 
lrwxrwxrwx 1 root root 29 Aug 28 14:31 /etc/pam.d/password-auth -> /etc/authselect/password-auth
```

What? I changed it!

```
$ sudo authselect select custom/mine
Profile "custom/mine" was selected.
The following nsswitch maps are overwritten by the profile:
- passwd
- group
- netgroup
- automount
- services

Make sure that SSSD service is configured and enabled. See SSSD documentation for more information.

$ sudo authselect select custom/mine
sudo: PAM account management error: Permission denied
sudo: a password is required
```

Aha! Getting somewhere. I guess you have to `authselect` again after any edits? I would have thought it would just change the symlinks. Weird.

Good job I opened that root terminal or I would have totally screwed my system. This is such a nice easy configuration change.

19. `# authselect select custom/mine` again a couple of times to be sure.

20. Try again. OMG it actually worked! I was half suspecting that I wouldn't get this far.

Good luck normal users doing this.

**Comment by thkukuk** [comment]
Since I did some timing measurements for a talk:
There is another security aspect to be noted: finding out if an account exists and thus if it is worth spending the time to try a brute force attack or not.
With the default 2 second delay (means a random delay between 2 and 3 seconds) it is impossible to find out through measurements, if the account is a valid one or not.
If you go down in the millisecond range, the random delay range also get's smaller and it's starting to become possible to find out if the account exits or not.

**Comment by Timmmm** [comment]
Can you explain how? I can't see why a 4x difference would make any more than a 4x difference in required measurements. 

Also if that *is* an issue then it should probably just be fixed properly, e.g. by running the delay in parallel with the auth check.

**Comment by thkukuk** [comment]
> I can't see why a 4x difference would make any more than a 4x difference in required measurements.

Sorry, that sentence doesn't make any sense to me.

If your variance gets smaller than the time needed for the operation it's useless, as you can guess the result of the operation and it's not hidden in the randomness.
And the current behavior is the correct fix, if the randomness is higher than the operation you cannot make any guess what it was.

**Comment by BenBE** [comment]
Configuring libpam is a PITA as it is. Every day that I don't have to touch anything even remotely near it is a [clear win in quality of life](https://m.xkcd.com/963/). So if you wonder why people take "two decades to even google if it is configurable" is a clear sign of the whole thing being a mess.

Frankly, "disable the delay in `pam_unix` and add `pam_faildelay`" is not a workaround so much as a crutch. Worse, it doesn’t guarantee your new setup behaves the same everywhere – subtly different module stacks might invoke `pam_faildelay` in one place but not another.

So just get that damn thing fixed in `pam_unix` to properly handle a configurable time instead of a binary on/off switch. And while at it, you can just update the default to 500ms which improves the interactive use case (most people are encountering this), while not impacting the non-interactive use-case (botnets¹ trying to hack your machine) much.

¹the cloud™ is cheap enough to just scale 4x if you really wanted to.

**Comment by Timmmm** [comment]
> If your variance gets smaller than the time needed for the operation it's useless, as you can guess the result of the operation and it's not hidden in the randomness.

Ah I see where we're miscommunicating... It's been a while since I checked the PAM source so I was assuming that the delay was *on top of* the authentication checks, whereas you were assuming it was *in parallel* with them.

1. If it is on top of them (in series) then you absolutely need randomised timing so that a large number of measurements are required to measure the underlying auth delay. In that case a 4x reduction in standard deviation means you need 4x fewer measurements to get the same accuracy on your auth delay (or maybe 16x; been a while since my statistics).
2. If it's in parallel then you technically don't need any randomness at all, so I'm not sure why it was added (`git blame` doesn't go back far enough unfortunately). It's probably a good idea just in case though. But as you say you need to know that your actual authentication process takes less time than the delay.

So what does PAM actually do? I checked the code. It has:

```
    if (pamh->former.choice == PAM_NOT_STACKED) {
	_pam_sanitize(pamh);
	_pam_start_timer(pamh);    /* we try to make the time for a failure
				      independent of the time it takes to
				      fail */
    }

    retval = _pam_dispatch(pamh, flags, PAM_AUTHENTICATE);

    if (retval != PAM_INCOMPLETE) {
	_pam_sanitize(pamh);
	_pam_await_timer(pamh, retval);   /* if unsuccessful then wait now */
	D(("exiting"));
    } else {
```

So it looks like it does it in parallel right? Not so fast....


```
void _pam_await_timer(pam_handle_t *pamh, int status)
{
    unsigned long long delay;
    D(("waiting?..."));

    delay = _pam_compute_delay(pamh->fail_delay.begin,
			       pamh->fail_delay.delay);
    if (pamh->fail_delay.delay_fn_ptr) {
       // This is only used if the app (e.g. sshd) sets a custom delay function.
```
  } else if (status != PAM_SUCCESS && pamh->fail_delay.set) {

	D(("will wait %llu usec", delay));

	if (delay > 0) {
	    struct timeval tval;

	    tval.tv_sec  = delay / 1000000;
	    tval.tv_usec = delay % 1000000;
	    select(0, NULL, NULL, NULL, &tval);
	}
    }
```

Ok great so `_pam_start_timer` sets `.begin` and then `_pam_compute_delay` will use that to compute a delay relative to it, right? Nope!

```
static unsigned long long _pam_compute_delay(unsigned int seed,
					     unsigned int base)
{
     int i;
     double sum;
     unsigned long long ans;

     for (sum=i=0; i<3; ++i) {
	  seed = _pam_rand(seed);
	  sum += (double) ((seed / 10) % 1000000);
     }
     sum = (sum/3.)/1e6 - .5;                      /* rescale */
     sum = base*(1.+sum);
     ans = sum > (double) ULLONG_MAX ? ULLONG_MAX : (unsigned long long) sum;
     D(("random number: base=%u -> ans=%llu\n", base, ans));

     return ans;
}
```

It uses it as a seed. So while it appears there was some intention to make the delay run in parallel, it actually doesn't. (Unless I've missed something, which is totally possible.)

I'm unsure if this is a 25 year old bug or not... but either way currently it means that reducing the variance in the delay doesn't add any extra risks (other than needing fewer measurements).

**Comment by CervEdin** [comment]
FWIW the 2s delay can be a major PITA. I frequently change keyboards & layouts and am often typing passwords double-blind. I can't see the keys I type (and naturally I don't see the keys registered either). The de-facto minimum delay of 2s is much longer than delays on Windows, OSX, Android, IOS, FreeBSD or pretty much any other system I interface with, and it makes typing my password on Linux much more annoying.

It's not my business to decide what default should be for everyone else. In the same spirit, I don't think others should decide what a sane default for me should be either. 2s seems like a reasonable default, permitted it can be changed. But not being able to change to a lower setting without doing some kind of occult ritual isn't what I would call configurable.

@Timmmm I think I got to something like step 14 last time I tried and was ready to give up until I found your comments. With the help of what you wrote, ChatGPT and `pamtester` I was finally able to change the default.

For future reference, I did something along these lines:

# Reducing PAM Authentication Delay with a Custom `authselect` Profile

This guide shows how to shorten the authentication failure delay on Fedora/RHEL
using a custom **authselect** profile.
We’ll configure:

- `pam_faildelay.so delay=500000` (0.5 s)
- `nodelay` for `pam_unix.so` on **auth** lines
- *(optional)* disable `pam_faillock`’s own delay if you use faillock

---

## 1️⃣ Check the Current Profile
```bash
sudo authselect current
```
Note the active profile (e.g. `sssd`).

---

## 2️⃣ Create a Custom Profile
```bash
sudo authselect create-profile myauth -b sssd
```

> This creates editable files under:
> `/etc/authselect/custom/myauth/`

---

## 3️⃣ Activate the Custom Profile
Enable the same features you used with the base profile:
```bash
sudo authselect select custom/myauth \
  with-silent-lastlog with-fingerprint with-mdns4
```

---

## 4️⃣ Edit the PAM Stacks

Open these files:
```bash
sudoedit /etc/authselect/custom/myauth/system-auth
sudoedit /etc/authselect/custom/myauth/password-auth
```

### In the **auth** section:
1. Make sure `pam_faildelay` is early:
   ```pam
   auth    required    pam_env.so
   auth    required    pam_faildelay.so delay=500000
   ```

2. Add `nodelay` to the `pam_unix.so` line:
   ```pam
   auth    sufficient  pam_unix.so ... nodelay
   ```

3. Leave `account`, `password`, and `session` sections as they are.

---

## 5️⃣ Apply Changes
```bash
sudo authselect apply-changes
```
(or re-run `authselect select` with the same arguments)

---

## 6️⃣ Verify the Effective Config
```bash
grep -nE 'pam_faildelay|pam_unix\.so|pam_faillock' \
  /etc/pam.d/system-auth /etc/pam.d/password-auth
```

You should see:

- `pam_faildelay.so delay=500000` before other auth modules
- `pam_unix.so … nodelay` on **auth** lines
- `pam_faillock.so` lines if you’re using faillock

---

## 7️⃣ Test Without Logging Out
Install **pamtester**:
```bash
sudo dnf install -y pamtester
```

Measure delay for a bad password:
```bash
time pamtester login "$USER" authenticate
```

> `real` should be about **0.5 s** instead of the default 2 s.

(You can also use `time su - "$USER"`.)

---

## 8️⃣ Troubleshooting
- If changes don’t show up under `/etc/pam.d/`, make sure you edited the files in
  `/etc/authselect/custom/myauth/` and ran `authselect apply-changes`.
- To revert:
  ```bash
  sudo authselect select sssd --force
  sudo authselect apply-changes
  ```

---

✅ Your system now uses a short, explicit failure delay via `pam_faildelay`,
with no extra sleeps from `pam_unix` or `pam_faillock` (if disabled).

**Comment by krukai** [comment]
I landed here after looking into why making a typo while entering my password in KDE's login manager lead to janky behavior, including the ~2 second delay before I could enter my password again.

I find the UX reasoning brought forward not too different from statements such as "people prefer having ad breaks because it gives them a chance to go to the bathroom". This sounds like subconsciously making the best of what is otherwise inconvenient or downright annoying; it's not like ad breaks were invented so people could use the bathroom. Regardless of what my perception is, some digging reveals findings such as this one from the field of human-computer interaction:

> The present findings demonstrate that even small delays in the performance of computers have   negative effects. From the perspective of the user our data suggests that eliminating even occasional small delays has the potential to improve work-satisfaction and well-being, which may in turn result in increased productivity and decreased stress.
_Szameitat, A. J., Rummel, J., Szameitat, D. P., & Sterr, A. (2009). Behavioral and emotional consequences of brief delays in human–computer interaction. International Journal of Human-Computer Studies, 67(7), 561-570._

This pretty much describes what this delay does to me, although one may argue that the experiment conducted in this study does not properly capture the scenario discussed here. Unfortunately, since other research suggests that users are more accepting of delays if there's a (seemingly) good reason, an experiment tailored around failed password attempts would be at risk to measure acceptance or perceived security, rather than actual preference. So, assuming findings such as the one quoted above aren't enough, what is? This question has not been sufficiently answered yet, as far as I can tell.

The discussion around the security aspect kinda fizzled out, but it seems like an exponential delay would be a good approach, if not a better one, assuming it applies in parallel as well. But does that even matter if the stance on the long-standing default is so rigid? After all, switching to an exponentially increasing delay by default implies lowering the initial one as well.

So, we are back at configuration changes, I suppose? Looking at the steps provided by other users however, I unfortunately find it borderline dismissive to claim it's configurable, assuming that's the best way to do it. Technically possible, sure, but arcane enough to most users to be wary of potential implications beyond the intended change. So if that's what it takes to change the default, then I am more inclined to live with the annoying delay than attempt to change it. That's not an expression of preference for the current default, but defeat.

It would be very helpful to know which path is worth pursuing, and what the acceptance criteria are. Perhaps it's actually much easier to configure than shown above, but I didn't see any comment stating so. Personally, I like the simplicity of lowering the default to 0.5 seconds. Thanks!
